### DISCUSSION TITLE
Asset download during nextjs export

### DISCUSSION THREAD

**Discussion by sandys** [vercel/next.js#22057] [Ideas]
# Feature request

A related conversation to this feature happened here https://github.com/zeit/next.js/issues/9524#issuecomment-574568879

if i use a headless CMS (even wordpress or graphcms or whatever), the asset url is used in the static html.

There are two preferences here - that asset links be used as such.
But more likely - download the asset, build the html (link to it locally) and then layer a CDN in front. This is the far more acceptable practice.

This also ties in very well with using deployment systems like Netlify - which have a far, far more suitable globally available infrastructure than something like DatoCMS or Graphcms. So if im using Netlify as deployment, i would want everything to be served from Netlify domain and let it work its magic.

We understand that this increases build times. This is ok for our usecase where we use a CMS just for the sake of usability, but deployment has to follow guidelines of compliance (geographic restriction of assets, etc) so we would like the entire build to be inside the export folder. This would mean that CMS that have CDN in front of it (e.g. datocms and www.datocms-assets.com which is Imgix) may not be ideal.

There are tons of similar requests. 

https://spectrum.chat/next-js/general/how-would-you-handle-importing-remote-images-on-nextjs-static-export~30b2ba84-bc27-4da7-9ec8-21e4d5d287a3

on gatsby side as well - gatsbyjs/gatsby#14076

https://spectrum.chat/gatsby-js/general/adding-remote-images-during-node-creation~e704e6fb-24b2-46c6-b1fc-93189d2e28a4

https://github.com/njosefbeck/gatsby-source-stripe/#downloading-files

**Comment by Timer** [comment]
This is covered by #9524

**Comment by sandys** [comment]
@Timer sorry for the late response. but this is NOT covered by #9524 . In fact, i was told there to open a new issue since this will not be considered for inclusion under SSG improvements.

Is it possible to do this using any kind of custom code, etc ? this is the only thing blocking us from moving entirely to Nextjs from Gatsby.

**Comment by erhankaradeniz** [comment]
@Timer I think @sandys is right here, I'm not seeing anything in that RFC which would solve this issue? Please correct me if I'm wrong, since I'd like to make use of this feature too since I switched to almost 100% SSG pages with the canary builds.

**Comment by sandys** [comment]
hi guys,
now that nextjs 9.3 is released with next-gen SSG support ( https://nextjs.org/blog/next-9-3#next-gen-static-site-generation-ssg-support ), im wondering if there is any way to achieve what was mentioned in this bug ?

**Comment by sandys** [comment]
@timer - any chance you can take a look at this and comment ? if there's a way to do this using the existing Nextjs framework, that would be awesome as well.

**Comment by franciscorubin** [comment]
Any news on this? Ideally this should just store images fetched during getStaticProps on .next/static/images like the rest of the images of the website. The fetching will only happen on build time (getStaticProps), so this shouldn't be a problem technically.

**Comment by mixtur** [comment]
I think it should also avoid duplicate downloads and allow post processing of whatever was downloaded afterwards.
So you mark something as an asset in getStaticProps and then in another magic export you do something with it. (Resize, compress etc)

**Comment by pavelbelyaev98** [comment]
Any news on this?

**Comment by Ecksters** [comment]
In my case my headless CMS is internal-only, and only accessible during the build process, currently I'm having to hack together a solution to copy the uploads folder from the headless CMS into my build export's `out` folder so that the URLs provided by the CMS can be accurate.

It'd be awesome, if NextJS could handle fetching these assets during the build for me, including potentially doing the image optimizations that are already available during the build.

**Comment by sandys** [comment]
Same issue for us. The CMS is private.

We would love this feature.

**Comment by le-martre** [comment]
This would really be a superb feature, I really need it as well :) Will check out if it's possible to hack something together but I imagine someone smarter than me would have figured it out by now :p

**Comment by sandys** [comment]
Unfortunately we couldn't figure it out. We had to move one of our projects
to Gatsby cos of this :(

Hoping that nextjs includes this.

**Comment by le-martre** [comment]
I actually managed to hack something together. Beware, it's as dirty as it gets, I just stopped when I got it working but didn't clean anything up and on top of that I suck at JS..

Basically, I just added a step after "next build && next export" that parses all the html files and finds all the images' srcs and srcsets. I filter them down to only those that start with specific strings (example "https://www.datocms-assets.com/"), I download everything then replace everything in the HTML and JSON files. 

It did work on Netlify as well first try.. Of course I STRONGLY encourage anybody that wants to use it to go over everything and refactor it because it really is just a Proof of Concept. But I figured it might help other people :) Cheers !

https://gist.github.com/MartinVandersteen/7235dbf42f601d83744067a72c4348fa

**Comment by robomatic** [comment]
only slightly less dirty ... I riffed on yours @MartinVandersteen.
https://gist.github.com/robomatic/8dccf289e33aac6ef92a1c4b6e69ca20

**Reply by le-martre** [reply]
Works really well ! Thanks a lot for that. 

For anyone using Datocms as well be careful that they seem to have appended the project number at the end of the URLS, so the url to replace might have become `https://www.datocms-assets.com/12345` for some of you (12345 being a project id)

**Reply by le-martre** [reply]
I had to update yours again so that it changes URLs in the JSON files as well. Not changing them caused weird issues where on the first load of some pages it would grab the datocms URL from the JSON file but would then use the local url on subsequent reloads.

https://gist.github.com/MartinVandersteen/7235dbf42f601d83744067a72c4348fa

**Comment by jens1101** [comment]
I recently had this same challenge and stumbled upon this discussion. Our private CMS returns to us URLs for each asset and we wanted to make these external assets form part of the NextJS build. I took a different approach from the suggestions listed here, so I decided that I wanted to share my solution.

**Step 1:** Add the following to `next.config.js`:
```JS
module.exports = {
  serverRuntimeConfig: {
    projectRoot: __dirname
  }
  // ...
};
```
This allows us to easily get the directory path of the repo later on.

**Step 2** In the `getStaticProps` for each page we fetch and parse the necessary data from the CMS. During the data parsing process we also have to parse each asset. As part of that step we download each asset to the public directory. I decided to download them to "public/uploads" for organisational purposes. This is also where the server runtime config comes in handy. The code looks something like this:
```js
const targetDirectory = path.join(
  serverRuntimeConfig.projectRoot,
  'public/uploads',
);

await downloadAsset({
  url,
  targetDirectory,
}),
```

The download function is fairly similar to [this SO answer](https://stackoverflow.com/a/62786397/1358008).

**Step 3:** Use the new path to the asset in the public directory. This is fairly simple since "/" refers to the public directory. Here's my code:
```js
return `/uploads/${path.basename(url)}`;
```

And that's it. Now the assets will be part of your build, since they are inside of the "public" folder.

**(Recommended) Step 4:** I recommend you add the asset target path to your `.gitignore` file so that the assets don't accidentally get checked into source control.
